tools/clang/lib/SPIRV/LowerTypeVisitor.cpp

```diff
@@ -852,5 +859,119 @@ LowerTypeVisitor::populateLayoutInformation(
   return result;
 }
 
+SpirvDebugTypeComposite *LowerTypeVisitor::lowerDebugTypeComposite(
+    const RecordType *structType, const SpirvType *type,
+    llvm::SmallVector<StructType::FieldInfo, 4> &fields) {
+  const auto *decl = structType->getDecl();
+  const SourceLocation &loc = decl->getLocStart();
+  const auto &sm = astContext.getSourceManager();
+  uint32_t line = sm.getPresumedLineNumber(loc);
+  uint32_t column = sm.getPresumedColumnNumber(loc);
+  StringRef linkageName = type->getName();
+
+  // TODO: Update linkageName using astContext.createMangleContext().
+  //
+  // Currently, the following code fails because it is not a
+  // FunctionDecl nor VarDecl. I guess we should mangle a RecordDecl
+  // as well.
+  //
+  // std::string s;
+  // llvm::raw_string_ostream stream(s);
+  // mangleCtx->mangleName(decl, stream);
+
+  uint32_t tag = 1;
+  if (decl->isStruct())
+    tag = 1;
+  else if (decl->isClass())
+    tag = 0;
+  else if (decl->isUnion())
+    tag = 2;
+  else
+    assert(!"DebugTypeComposite must be a struct, class, or union.");
+
+  bool isPrivate = decl->isModulePrivate();
+
+  std::string name = type->getName();
+
+  // TODO: Update parent, size, and flags information correctly.
+  RichDebugInfo *debugInfo = &spvContext.getDebugInfo().begin()->second;
+  const char *file = sm.getPresumedLoc(loc).getFilename();
+  if (file)
+    debugInfo = &spvContext.getDebugInfo()[file];
+  auto *dbgTyComposite =
+      dyn_cast<SpirvDebugTypeComposite>(spvContext.getDebugTypeComposite(
+          type, name, debugInfo->source, line, column,
+          /* parent */ debugInfo->compilationUnit, linkageName,
+          /* size */ 0,
+          /* flags */ isPrivate ? 2u : 3u, tag));
+
+  // If we already visited this composite type and its members,
+  // we should skip it.
```

Comment:
If the struct is actually an empty struct with no members, `members.empty()` will be `true` even if we have visited it in the past. Therefore, I think the following lines [914-916] should be moved before this `if` condition.

Reply:
Hmm .. This code looks fine. If it is an empty struct with no members, `members.empty()` will be `true` and it will not return in line 912. Just minor issue is the line [914-916] will be executed twice if we have visited it in the past.

I added a unit test but it looks fine.